fctesting/sandbox/fc-latest

```diff
@@ -0,0 +1,68 @@
+FROM ubuntu:18.04
+FROM golang:1.11
+
+ADD . /firecracker-go-sdk/
+
+ARG TMP_BUILD_DIR=/tmp/build
+ARG FIRECRACKER_SRC_DIR="/firecracker"
+ARG FIRECRACKER_BUILD_DIR="$FIRECRACKER_SRC_DIR/build"
+ARG CARGO_REGISTRY_DIR="$FIRECRACKER_BUILD_DIR/cargo_registry"
+ARG FIRECRACKER_GO_SDK_SRC_DIR="/firecracker-go-sdk"
+ARG FIRECRACKER_VERSION="0.15.0"
+
+ENV CARGO_HOME=/usr/local/rust
+ENV RUSTUP_HOME=/usr/local/rust
+ENV PATH="$CARGO_HOME/bin:$PATH"
+ENV CARGO_TARGET_DIR="$FIRECRACKER_BUILD_DIR/cargo_target"
+ENV FC_TEST_DATA_PATH="$TMP_BUILD_DIR"
+
+# Install system dependecies
+#
+RUN apt-get update \
+    && apt-get -y install --no-install-recommends \
+        binutils-dev \
+        cmake \
+        curl \
+        file \
+        g++ \
+        gcc \
+        git \
+        iperf3 \
+        iproute2 \
+        jq \
+        libcurl4-openssl-dev \
+        libdw-dev \
+        libiberty-dev \
+        libssl-dev \
+        lsof \
+        make \
+        musl-tools \
+        net-tools \
+        openssh-client \
+        pkgconf \
+        zlib1g-dev \
+    && rm -rf /var/lib/apt/lists/*
+
+# Install the Rust toolchain
+#
+# Install 1.32.0 due to Firecracker crashing with 148 code
+RUN mkdir "$TMP_BUILD_DIR" \
+    && curl https://sh.rustup.rs -sSf | sh -s -- -y \
+    	&& rustup install 1.32.0 \
+	&& rustup override set 1.32.0 \
+        && rustup target add x86_64-unknown-linux-musl \
+    && cd /
```

Comment:
I suggest utilizing docker's [multi-stage](https://docs.docker.com/develop/develop-images/multistage-build/) builds here. I think this will allow to drop rust installation and significantly reduce apt-get install deps.

Something like:
```Dockerfile
FROM rust:latest as fc-build

# Build FC
RUN cargo build --release --features vsock --target-dir "$TMP_BUILD_DIR"
...

FROM go:1.12
...
COPY --from  fc-build firecracker-binary . # Copy FC binary from previous stage
...
```

Reply:
Does the `rust:latest` have musl with it? And we don't want latest as it doesn't work with firecracker, we want `1.32`.

Reply:
I'm not sure about `musl`.
As for versions, you can use `rust:1.32`

Reply:
Yea, that's why rustup was used to get musl. rustup will automatically grab rust and the like, so I don't think adding the `FROM rust:1.32` will save anything, cause we'd still need to grab rustup and install musl and rebuild.

Reply:
The rustup commands from the existing dockerfile will install musl and 1.32 as the defaults.

Reply:
> cause we'd still need to grab rustup and install musl

`rust:1.32` [already contains](https://github.com/rust-lang/docker-rust/blob/97a72441aad0ec26edf5175382c3d0022c84ba7d/Dockerfile-debian.template#L19) `rustc` and `rustup`
`rustup target add x86_64-unknown-linux-musl` adds `rust-std` component.
<details>

% rustup -v target add x86_64-unknown-linux-musl   
verbose: read metadata version: '12'
info: downloading component 'rust-std' for 'x86_64-unknown-linux-musl'
verbose: downloading file from: 'https://static.rust-lang.org/dist/2019-01-17/rust-std-1.32.0-x86_64-unknown-linux-musl.tar.xz'
verbose: downloading with curl
verbose: checksum passed
info: installing component 'rust-std' for 'x86_64-unknown-linux-musl'
verbose: creating temp directory: ...

</details>

Reply:
@mxpv - Ah cool! Will use that then :)

Reply:
Ah, so apparently the multistage builds are two separate containers. I need them to be the same since the binaries from the first stage needs to be in the build directory for the second stage. So, Im going to keep this as is.

Reply:
This is where you use `COPY --from prev_stage_name . . ` to transfer artifacts between stages.

Reply:
Okay, I separated the steps out which I think is nicer. However, I still needed to build the toolchain since no toolchain was provided ><